File: euler94/euler94.py
import math
from tqdm import tqdm


PERFECT_SQUARES = set()
for n in tqdm(range(1_000_000_000 // 3)):
    PERFECT_SQUARES.add(n**2)

# print(len(PERFECT_SQUARES))
# print(sorted(PERFECT_SQUARES, key=lambda x: abs(1643897025 - x))[:10000])

assert 1643897025 in PERFECT_SQUARES

def area_is_integer(a, b):
    h_squared = a**2 - (b//2)**2
    # print(h_squared)
    if math.sqrt(h_squared).is_integer(): print(h_squared)
    return h_squared in PERFECT_SQUARES


assert area_is_integer(17, 16)

# In order for area to be an integer, (b/2, h, a) must form a
# pythagorean triple. Thus, there is a correspondence between
#  - pythagorean triples of the form (k, _, 2k±1) and
#  - triangles with sides b = 2k and a = 2k±1
# 
# However, a billion isn't actually that big so we can brute force
# it by checking all odd a values and corresponding even b values :)


MAX_PERIMETER = 1_000_000_000
# MAX_PERIMETER = 100

total = 0
for a in tqdm(range(3, MAX_PERIMETER // 3, 2)):
    for b in (a - 1, a + 1):
        if area_is_integer(a, b):
            print(a, a, b)
            total += a + a + b

print(total)

File: euler94/euler94.cpp
#include <iostream>
#include <cmath>
#include <assert.h>


bool area_is_integer(long long a, long long b) {
    long long val = 4*a*a - b*b;
    long long root = sqrt(val);
    return root * root == val;
}




void euler94() {
    assert(area_is_integer(5, 6));

    const long long perimter_max = 1000000000;    // 1 billion

    long long total = 0;

    for (long long a = 3; a < perimter_max/3; a++) {
        for (long long b : {a - 1, a + 1}) {
            if (area_is_integer(a, b)) {
                // std::cout << a << " " << b << std::endl;
                total += a + a + b;
            }
        }
    }

    std::cout << total << std::endl;

}


#include "../helpers/timing.cpp"
int main() {
    auto duration = func_time_ms(euler94);
    std::cout << "\n(" << duration << " ms)" << std::endl;
}

// 518408346
